lib/tests/formulae_dependents.rb

```diff
@@ -171,17 +171,18 @@ def install_dependent(dependent, testable_dependents, args:, build_from_source:
         if testable_dependents.include? dependent
           test "brew", "install", "--only-dependencies", "--include-test", dependent.full_name
 
-          dependent.recursive_dependencies.each do |dependency|
-            next if dependency.build? && !dependency.test?
-
+          # Traverse the dependency tree to check for formulae we need to link
+          dependencies_to_link = dependent.recursive_dependencies do |dep_dependent, dependency|
+            Dependency.prune if dependency.build? && !dependency.test?
+            Dependency.prune if dependency.test? && dep_dependent != dependent
             dependency_f = dependency.to_formula
-            # We don't want to attempt to link runtime deps of build deps.
-            next unless dependency_f.any_version_installed?
-            next if dependency_f.keg_only?
-            next if dependency_f.linked_keg.exist?
+            Dependency.skip if dependency_f.keg_only?
+            Dependency.skip if dependency_f.linked_keg.exist?
+          end.map(&:to_formula)
 
-            unlink_conflicts dependency_f
-            test "brew", "link", dependency_f.full_name
+          dependencies_to_link.each do |f|
+            unlink_conflicts f
+            test "brew", "link", f.full_name
```

Comment:
Seems easier to put all the dependency selection logic in one place, no?

Reply:
I suppose we could move at least the `linked?` check here and then the cache key can be just `"test-bot-link"` since the filtering should no longer be dynamic.

Reply:
My thought process is that this would make the code less complex because by calling `to_formula` only once (but maybe I'm wrong).

Reply:
You do raise a point since `to_formula` is not that cheap.

Reply:
Hmmm ok fair enough